More descriptive function information

JaroslavTulach opened the pull request:
### Pull Request Description

Fixes #7359 by printing more information about the function including partially applied arguments and over-saturated arguments.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
- All code has been tested:
  - [x] Unit tests have been written where possible.

JaroslavTulach committed:
More descriptive function information

JaroslavTulach commented on engine/runtime/src/test/java/org/enso/interpreter/test/SignatureTest.java:
```diff
@@ -353,6 +354,73 @@ public void binaryWithVec() throws Exception {
+    try {
+      var res = mix.execute(7);
+      fail("No result expected: " + res);
+    } catch (PolyglotException ex) {
```
After 8014806e6d046bc53fafe20936423cf49bb300d0 it prints error:
```
Type error: expected `x` to be V, but got V.Val[partial:4-4] a=1 b=7 c=_.
```

JaroslavTulach commented on engine/runtime/src/test/java/org/enso/interpreter/test/SignatureTest.java:
```diff
@@ -353,6 +354,73 @@ public void binaryWithVec() throws Exception {
+    try {
+      var res = mix.execute(7);
+      fail("No result expected: " + res);
+    } catch (PolyglotException ex) {
```
Prints following error:
```
Type error: expected `x` to be Integer, but got oversaturated.fn[oversaturated:3-6] self=oversaturated a=2 b=3 +a=4.
```
The `+a` means over-saturated `a` argument.

**Update**: After 8014806e6d046bc53fafe20936423cf49bb300d0 it also includes missing `c=_` argument.

radeusgd commented:
This looks really awesome and will be immensely helpful! 🎉 

I think we should still add some additional comment in case the RHS is a Function or constructor, to be more friendly to less advanced users. Perhaps [taking inspiration from `Index_Sub_Range` special handling](https://github.com/enso-org/enso/blob/d078a37a6a98022e8374b3273ec078982ecaccd8/distribution/lib/Standard/Base/0.0.0-dev/src/Data/Index_Sub_Range.enso#L283-L292) we could hint to the user that the function/constructor is likely missing some arguments?

Maybe also if over-saturated arguments show up we should explain it? I don't think anyone seeing this error message will be able to dig up this PR to figure out what `+` means there.

This is a potentially complicated error for beginners so I think it's better for it to be more verbose than to be difficult to understand.

--------

If possible, it would also be cool to add a test to the Enso-side tests checking this in a 'practical scenario'.

Perhaps:
```diff
Index: test/Table_Tests/src/Common_Table_Operations/Join/Join_Spec.enso
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/Table_Tests/src/Common_Table_Operations/Join/Join_Spec.enso b/test/Table_Tests/src/Common_Table_Operations/Join/Join_Spec.enso
--- a/test/Table_Tests/src/Common_Table_Operations/Join/Join_Spec.enso	(revision 72031726a6181887f62691c194161199382958eb)
+++ b/test/Table_Tests/src/Common_Table_Operations/Join/Join_Spec.enso	(date 1692717513911)
@@ -377,6 +377,13 @@
                 problems3 = [Floating_Point_Equality.Error "Z", Floating_Point_Equality.Error "X"]
                 Problems.test_problem_handling action3 problems3 tester3
 
+        Test.specify "should report a user-friendly type error if the constructor is missing some arguments" <|
+            t1 = table_builder [["Text", ["a", "b"]]]
+            r1 = Panic.recover Any (t1.join t1 on=[Join_Condition.Equals] on_problems=Problem_Behavior.Ignore)
+            r1.should_fail_with Type_Error
+            r1.catch.to_display_text . should_contain "missing argument"
+            r1.catch.to_display_text . should_contain "Join_Condition.Equals"
+
         Test.specify "should correctly handle nulls in equality conditions" pending=db_todo <|
             t1 = table_builder [["X", ["A", Nothing, "a", Nothing, "ą"]], ["Y", [0, 1, 2, 3, 4]]]
             t2 = table_builder [["X", ["a", Nothing, Nothing]], ["Z", [10, 20, 30]]]
```

Akirathan commented:
Very nice. I have another suggestion. What about displaying unapplied arguments as well? The behavior on this PR:
```
my_func x y =
  x + y

main = my_func
```
Prints
```
tmp.my_func[tmp.enso:1-2] self=tmp
```

With the following patch:
```diff
diff --git a/engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/function/Function.java b/engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/function/Function.java
index 8516bd4b4..025ed0961 100644
--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/function/Function.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/function/Function.java
@@ -423,9 +423,11 @@ public final class Function implements EnsoObject {
         .append(end)
         .append("]");
     for (var i = 0; i < schema.getArgumentsCount(); i++) {
+      sb.append(" ").append(schema.getArgumentInfos()[i].getName()).append("=");
       if (preAppliedArguments != null && preAppliedArguments[i] != null) {
-        sb.append(" ").append(schema.getArgumentInfos()[i].getName()).append("=");
         sb.append(iop.toDisplayString(preAppliedArguments[i], false));
+      } else {
+        sb.append("_");
       }
     }
     if (schema.getOversaturatedArguments() != null) {
```

It prints
```
tmp.my_func[tmp.enso:1-2] self=tmp x=_ y=_
```
To visualize that `x` and `y` arguments were not provided. The `_` placeholder is just an idea, it could be different. What do you think @radeusgd @JaroslavTulach ?

Commit:
Using _ to visualize not provided function arguments

JaroslavTulach committed:
Check unapplied arguments are printed in the function description

JaroslavTulach commented:
> It prints
> 
> ```
> tmp.my_func[tmp.enso:1-2] self=tmp x=_ y=_
> ```
Good idea to use `_`. I've just integrated your patch as 8014806

JaroslavTulach committed:
Include column range for single line functions

JaroslavTulach committed:
Check behavior of suspended arguments

JaroslavTulach commented:
> it would also be cool to add a test to the Enso-side tests checking this in a 'practical scenario'.

Probably better to add such integration tests when we agree on the format of error message.

I tried to apply your patch, but it is not compilable (missing import), neither working (after fixing the import). Yields:
```
   - [FAILED] should report a user-friendly type error if the constructor is missing some arguments [7ms]
        Reason: Expected error Type_Error, but error (Inexhaustive_Pattern_Match.Error Constructor<Equals>) has been returned (at /home/devel/NetBeansProjects/enso/enso.master/test/Table_Tests/src/Common_Table_Operations/Join/Join_Spec.enso:384:13-42).
```
If you get your test working, feel free to commit. Easier and more reliable than sending _"sketch only"_ not fully working patches.

JaroslavTulach commented on engine/runtime/src/test/java/org/enso/interpreter/test/SignatureTest.java:
```diff
@@ -353,6 +354,109 @@ public void binaryWithVec() throws Exception {
+      var res = mix.execute(0);
+      fail("No result expected: " + res);
+    } catch (PolyglotException ex) {
+      assertContains("Type error", ex.getMessage());
```
Yields following single line message:
```
Type error: expected `x` to be Integer, but got 
suspended.fn[suspended:3-5] 
self=suspended 
a=_ 
b=suspended.mix<arg-b>[suspended:9:28-30] 
c=suspended.mix<arg-c>[suspended:9:20-22].
```
I assume we want _single line_ messages although the description of arguments may be pretty looong.

JaroslavTulach committed:
Try to apply ... arguments hint

JaroslavTulach commented:
> we could hint to the user that the function/constructor is likely missing some arguments?

Done in 4105f51b05

JaroslavTulach committed:
Additional test also with over-saturated argument

radeusgd commented on test/Tests/src/Semantic/Error_Spec.enso:
```diff
@@ -341,4 +342,23 @@ spec =
+        Test.specify "try to apply one argument" <|
+            r = Panic.recover Type_Error <| neg (my_func -5)
+            r . should_fail_with Type_Error
+            r.to_display_text.contains "Try to apply y argument" . should_be_true
```
This will yield nicer message if it ever failed :)
```suggestion
            r.to_display_text.should_contain "Try to apply y argument."
```

radeusgd commented on test/Tests/src/Semantic/Error_Spec.enso:
```diff
@@ -341,4 +342,23 @@ spec =
+        Test.specify "try to apply two arguments" <|
+            r = Panic.recover Type_Error <| neg my_func
+            r . should_fail_with Type_Error
+            r.to_display_text.contains "Try to apply x, y arguments" . should_be_true
```
```suggestion
            r.to_display_text.should_contain "Try to apply x, y arguments."
```

radeusgd committed:
proposing 'integration' test

Current state:
Type Errors:  [4/5, 452ms]
    - everything is ok [1ms]
    - try to apply one argument [253ms]
    - try to apply two arguments [62ms]
    - [FAILED] report unapplied constructor nicely [77ms]
        Reason: The value (Error: Type error: expected `x` to be My_Type, but got Constructor<My_Type.Value>.) did not contain the element (Try to apply foo argument.) (at C:\NBO\enso\test\Tests\src\Semantic\Error_Spec.enso:370:13-73).
    - report partially applied constructor nicely [56ms]

radeusgd commented:
I pushed a commit ([5c2ef16](https://github.com/enso-org/enso/pull/7629/commits/5c2ef16c42b7efe52f48d968258ba091211ba441)) with a proposed test.

It is currently failing one case:
```
Type Errors:  [4/5, 452ms]
    - everything is ok [1ms]
    - try to apply one argument [253ms]
    - try to apply two arguments [62ms]
    - [FAILED] report unapplied constructor nicely [77ms]
        Reason: The value (Error: Type error: expected `x` to be My_Type, but got Constructor<My_Type.Value>.) did not contain the element (Try to apply foo argument.) (at C:\NBO\enso\test\Tests\src\Semantic\Error_Spec.enso:370:13-73).
    - report partially applied constructor nicely [56ms]
```

Do you think we can extract the argument names from the constructor? (we may need some special case for this, since normally when calling `to_text` on a constructor, we probably don't want them, we'd probably only want them when displayed in the `Type_Error`).

If this is not implementable, we can change the test a bit, but ideally I think it would be great if it could work like this. Note that once one argument in the ctor is applied, it acts as a function and the missing arguments _do show up_ (the latter test case that is passing). And IMO a ctor with 0 vs 1 arguments applied does not differ that much so it shouldn't probably behave much differently in this error - which these tests try to enforce.

JaroslavTulach committed:
Merge branch 'wip/jtulach/FnTypeError_7359' of enso:enso-org/enso into wip/jtulach/FnTypeError_7359

JaroslavTulach committed:
Avoid stackoverflow in Atom.toString by not printing the self argument of to_text function

JaroslavTulach committed:
Include list of pending fields in a atom constructor textual representation

JaroslavTulach committed:
Avoid requesting to provide defaulted arguments

JaroslavTulach commented on test/Tests/src/Semantic/Error_Spec.enso:
```diff
@@ -341,4 +344,66 @@ spec =
+        Test.specify "report unapplied constructor with default value nicely" <|
+            r = Panic.recover Type_Error <| extract My_Type.Default_Value
+            r . should_fail_with Type_Error
+            r.to_display_text.should_contain "Try to apply bar argument."
```
This message is certainly correct. 
```ruby
type My_Type
    Default_Value foo=2 bar
```
We don't want to encourage users to specify defaulted arguments with the _"Try to apply..."_ message.

However we don't include the defaulted arguments in display of functions neither atom constructors at all. Missing such a slick `x=_` syntax. Moreover default value is specified as AST and I don't want to evaluate it...

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/AtomConstructor.java:
```diff
@@ -318,7 +318,14 @@ Atom instantiate(Object... arguments) throws ArityException {
+    sb.append("Constructor<").append(getDisplayName()).append(">");
+    for (var f : getFields()) {
+      if (!f.hasDefaultValue()) {
+        sb.append(" ").append(f.getName()).append("=_");
```
This is a fix for [Radek's test](https://github.com/enso-org/enso/pull/7629#issuecomment-1689700450). Adds enough `field=_` so the `Type_Error` can pick it up and assemble proper _"Try..."_ message.

JaroslavTulach committed:
Convert obj toString behind @TruffleBoundary

mergify[bot] committed:
Merge branch 'develop' into wip/jtulach/FnTypeError_7359

hubertp commented on engine/runtime/src/test/java/org/enso/interpreter/test/SignatureTest.java:
```diff
@@ -353,6 +354,109 @@ public void binaryWithVec() throws Exception {
+      var res = mix.execute(7);
+      fail("No result expected: " + res);
+    } catch (PolyglotException ex) {
+      assertContains("Type error", ex.getMessage());
+      assertContains("expected `x` to be V", ex.getMessage());
+      assertContains("got V.Val[partial", ex.getMessage());
+      assertContains("a=1", ex.getMessage());
+      assertContains("b=7", ex.getMessage());
+      assertContains("c=_", ex.getMessage());
```
can't we just make it a single line? This feels a bit old to split it into numerous "contains" asserts. Same comment in other cases.

hubertp reviewed (approved):
minor nit re test definition

mergify[bot] committed:
Merge branch 'develop' into wip/jtulach/FnTypeError_7359

mergify[bot] committed:
Merge branch 'develop' into wip/jtulach/FnTypeError_7359
